File: csharp/tree-building/TreeBuilding.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class TreeBuildingRecord
{
    public int ParentId { get; set; }
    public int RecordId { get; set; }
}

public class Tree
{
    public int Id { get; set; }
    public int ParentId { get; set; }

    public List<Tree> Children { get; set; }

    public bool IsLeaf => Children.Count == 0;
}

public static class TreeBuilder
{
    public static Tree BuildTree(IEnumerable<TreeBuildingRecord> records)
    {
        if (records.Count() == 0) throw new ArgumentException("empty records");     
        
        var ordered = new SortedList<int, TreeBuildingRecord>();
        foreach (var record in records)
        {
            ordered.Add(record.RecordId, record);
        }
        if (ordered.Keys.First() != 0)                throw new ArgumentException("missing root");
        if (ordered.Keys.Last() != ordered.Count - 1) throw new ArgumentException("non-continuous list");
        
        records = ordered.Values;
        var trees = new List<Tree>();

        foreach (var record in records)
        {
            if (record.RecordId == 0 && record.ParentId != 0) throw new ArgumentException("invalid root");
            if (record.RecordId != 0 && record.ParentId >= record.RecordId) throw new ArgumentException("non-root with invalid parent");

            Tree t = new Tree { Id = record.RecordId, ParentId = record.ParentId, Children = new List<Tree>() };
            if (record.RecordId != 0)
            {
                var parent = trees.First(tree => tree.Id == record.ParentId);
                parent.Children.Add(t);
            }
            trees.Add(t);
        }
        return trees.First(t => t.Id == 0);
    }
}

File: fsharp/tree-building/TreeBuilding.fs
// This is the file you need to modify for your own solution.
// The unit tests will use this code, and it will be used by the benchmark tests
// for the "Mine" row of the summary table.

// Remember to not only run the unit tests for this exercise, but also the
// benchmark tests using `dotnet run -c Release`.
// Please refer to the instructions for more information about the benchmark tests.

module TreeBuilding

open TreeBuildingTypes

type Tree =
    | Leaf of int
    | Branch of int * Tree list

let recordId = function
    | Leaf id -> id
    | Branch (id, _) -> id
    

let isBranch = function 
    | Branch _ -> true
    | Leaf _ -> false

let children = function 
    | Branch (_, c) -> c
    | Leaf _ -> []

let buildTree records =
    if List.isEmpty records then failwith "Empty input"

    let checkRecordTuple (p, r) = 
        if r > 0 then 
            if p >= r then failwith "Non-root record with invalid parent"
            (p, r)
        else if r = 0 then 
            if p <> 0 then failwith "Root node is invalid" 
            (p, r)
        else failwith "Node with negative record ID"

    let checkIndex (i, (p, r)) = 
        if i <> r then failwith "Non-continuous list" 
        (p, r)

    let parentChildrenMap =
        records 
        |> List.map (fun r -> (r.ParentId, r.RecordId) ) 
        |> List.sortBy snd
        |> List.map checkRecordTuple
        |> List.indexed 
        |> List.map checkIndex
        |> List.tail
        |> List.groupBy fst 
        |> List.map (fun (x, y) -> (x, List.map snd y))
        |> Map.ofSeq 

    let rec buildHelper parent =
        match Map.tryFind parent parentChildrenMap with
        | Some children -> Branch (parent, children |> List.map buildHelper)
        | _             -> Leaf parent

    buildHelper 0